feat: next level of the datastore

dignifiedquire opened the pull request.

dignifiedquire committed:
feat: next level of the datastore

dignifiedquire committed:
chore: add license

dignifiedquire committed:
feat(key): use buffers as underlying data source

dignifiedquire committed:
fix(sharding): raw ops for fs backend

dignifiedquire committed:
docs(readme): updates

daviddias commented on README.md:
```diff
@@ -1,92 +1,93 @@
 
 ## License
 
-MIT
```
The README needs a reference to https://github.com/ipfs/go-datastore and also, list out all the methods that are not implemented, if any.

daviddias commented on flow-typed/npm/pull-stream_vx.x.x.js:
```diff
@@ -0,0 +1,367 @@
+}
+declare module 'pull-stream/util/tester.js' {
+  declare module.exports: $Exports<'pull-stream/util/tester'>;
+}
```
Will every module that uses flow have to generate this? Should there be a `flow-pull-stream` package instead?

dignifiedquire replied:
We should crate a definition and pr it to the flow-typed repo if we start using flow regularly yes, for now this is just a generated stub

daviddias replied:
Well, we can do that right now. pull-stream is part of the entire npm ecosystem, someone else will find it useful and it will save us space and time for when we decide to use it in other places.

dignifiedquire replied:
I don't have the time right now to do it, we can create an issue for it

daviddias commented on package.json:
```diff
@@ -1,30 +1,49 @@
-  "description": "datastore",
-  "main": "index.js",
+  "name": "interface-datastore",
+  "version": "1.0.0",
```
Let's keep at 0.1.0 until we have it all merged

daviddias commented on package.json:
```diff
@@ -1,30 +1,49 @@
+    "datastore"
   ],
-  "author": "Juan Benet <juan@benet.ai> (http://juan.benet.ai/)",
+  "author": "Friedel Ziegelmayer <dignifiedquire@gmail.com>",
```
Well, Juan kind of created the whole thing.

Mind adding yourself as contributor?

Note, I don't think these things should be important.

daviddias commented on src/index.js:
```diff
@@ -0,0 +1,75 @@
+
+export type Order<Value> = (QueryResult<Value>, Callback<QueryResult<Value>>) => void
+
+*/
```
Can all of this flow stuff be declared in a separate file?

dignifiedquire replied:
I'd rather not, otherwise I have to change the type imports everywhere

daviddias commented on src/key.js:
```diff
@@ -0,0 +1,314 @@
+ * - `new Key('/Comedy/MontyPython/Sketch:CheeseShop')`
+ * - `new Key('/Comedy/MontyPython/Sketch:CheeseShop/Character:Mousebender')`
+ *
+ */
```
This needs to go into the README

dignifiedquire replied:
this is jsdoc, will go into the generated api docs, as we do with other api docs

daviddias replied:
I meant, the README needs to have more info about its API :)

daviddias commented on src/shard-readme.js:
```diff
@@ -0,0 +1,26 @@
+with 'SC' being the last-to-next two characters and the 'B' at the
+beginning of the CIDv1 string is the multibase prefix that is not
+stored in the filename.
+`
```
Why is the readme a module?

dignifiedquire replied:
Because it needs to be included in the code. it's written into the directories that use sharding

daviddias replied:
Interesting

daviddias reviewed (changes requested):
This module still has the different implementations inside it. But you also extracted some of them, let's extract all of them, making this module as light as possible.

dignifiedquire commented:
> This module still has the different implementations inside it. But you also extracted some of them, can we just start off by extracting all of them?

I extracted all backing implementations, I don't think it makes sense to put all these wrapping implementations into their own repos right now, it just makes things even more complex :(

dignifiedquire committed:
chore: address some of the cr

dignifiedquire committed:
chore: remove pull-stream stub for now

dignifiedquire commented:
Ref: https://github.com/ipfs/js-ipfs/issues/787

daviddias commented:
> I extracted all backing implementations, I don't think it makes sense to put all these wrapping implementations into their own repos right now, it just makes things even more complex :(

Understood, that is because datastore has a lot of functionality, more than just using some storage backend. It seems that to make this clear we need

- interface-datastore - interface definition and tests
- js-datastore - implementation that has all of the functionalities
- js-datastore-fs and js-datastore-level - the storage drivers that js-datastore can mount

dignifiedquire commented:
Given that `datastore` is not available on npm, I would suggest `js-datastore-core` for the implementations, as `datastore-core` is available.

daviddias commented:
@dignifiedquire go for it :)

dignifiedquire committed:
feat: implement namespace store

dignifiedquire committed:
extract wrapper implementations into js-datastore-core

dignifiedquire commented:
@diasdavid extraction is done

dignifiedquire committed:
docs(readme): expand api docs

dignifiedquire commented:
@diasdavid updated readme to have more examples and api docs.

daviddias reviewed (changes requested):
The README is using flow notations, need to use our standard way of presenting the different types.

Other than that. LGTM

dignifiedquire committed:
docs(readme): use other method of describing types
